Syndication/AzureStack.MarketplaceSyndication.psm1

```diff
@@ -90,165 +83,1086 @@ function Sync-AzSOfflineMarketplaceItem {
+
+    if($notPresent)
+    {
+        Set-AzureStorageBlobContent -File $file.FullName `
+            -Container $syndicationContainerName `
+            -Blob $blobName `
+            -Context $ctx `
+            -Force | Out-Null
+
+        $fileURI = (Get-AzureStorageBlob -blob $blobName -Container $syndicationContainerName -Context $ctx).ICloudBlob.Uri.AbsoluteUri
+    }
+    else
+    {
+        Write-Host "$blobName exist in storage, skip upload"
+        $fileURI = $blobInfo.ICloudBlob.uri.AbsoluteUri
+    }
+
+    return $fileURI
+}
+
+function Ensure-SuccessStatusCode {
+    param(
+        [Parameter(Mandatory = $true)]
+        [System.Net.HttpStatusCode] $StatusCode
+    )
+
+    $ErrorActionPreference = 'Stop'
+
+    if (-not (Test-SuccessStatusCode -StatusCode $StatusCode)) {
+        throw "HTTP response status code is not successful: $StatusCode"
+    }
+}
+
+function Wait-AzsAsyncOperation {
+    param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [Uri] $AsyncOperationStatusUri,
+
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNull()]
+        [object] [ref]$Headers,
+
+        [Parameter(Mandatory = $false)]
+        [PSCredential] $AzsCredential
+    )
+
+    $ErrorActionPreference = 'Stop'
+
+    # max wait for two hours, otherwise treat it as failed
+    $currentAttempt = 0
+    $maxAttempts = 720
+    $sleepSeconds = 10
+
+    $stopwatch = [System.Diagnostics.Stopwatch]::StartNew()
+
+    while ($true) {
+        $response = InvokeWebRequest -Method GET -Uri $AsyncOperationStatusUri -ArmEndpoint $armEndpoint -Headers ([ref]$Headers) -MaxRetry 10 -AzsCredential $AzsCredential
+
+        Ensure-SuccessStatusCode -StatusCode $response.StatusCode
+
+        $operationResult = $response.Content | ConvertFrom-Json
+
+        if (Test-OperationResultTerminalState $operationResult.status) {
+            if ($operationResult.status -eq 'Succeeded') {
+                return $true
+            }
+
+            return $false
         }
 
-        else {
-            $a.popup("Legal Terms not accepted, canceling")
+        $currentAttempt++
+        if ($currentAttempt -ge $maxAttempts)
+        {
+            throw "Async operation was not finished after $currentAttempt retries. Provisiong state: $operationResult.status"
         }
 
+        Write-Debug "Sleeping for $sleepSeconds seconds, waiting time: $($stopwatch.Elapsed)"
+
+        Start-Sleep -Seconds $sleepSeconds
     }
 }
 
+function Test-OperationResultTerminalState {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $Value
+    )
+
+    return $Value -in @('Canceled', 'Failed', 'Succeeded')
+}
+
+function Test-SuccessStatusCode {
+    param(
+        [Parameter(Mandatory = $true)]
+        [System.Net.HttpStatusCode] $StatusCode
+    )
+
+    return [int]$StatusCode -ge 200 -and [int]$StatusCode -le 299
+}
 
+function InvokeWebRequest {
```

Comment:
this is pretending to be invoke web request for a generic URI but then assumes something about calling ARM and overwriting an auth header on any failure from the remote endpoint. Consider having a single function for calling an ARM URI which abstracts the auth context. Why are we not using Azure PowerShell for these operations?

Reply:
Will look at this later